### PR TITLE
Table: Adds new footer feature that can show Totals or other column calculations 

### PR DESCRIPTION
**Pull request by scottlepp**
**What this PR does / why we need it**:
Allows showing calculated values on a footer

- [x] show/hide footer toggle
- [x] multi field selection
- [x] one reducer selection
- [x] display reducer function name in first column (if column is not numeric)

image

### CODE REVIEW

**Review by torkelo — COMMENTED** [review]
This looks interesting!

File: public/app/plugins/panel/table/TablePanel.tsx [hunk 1]
```diff
@@ -91,10 +114,216 @@ export class TablePanel extends Component<Props> {
...
+
+    if (options.footerMode === 'none') {
+      return { cells: undefined };
+    }
+
+    if (options.footerMode === 'frame') {
+      const summaryFrame = data.series.find((f) => f.name === options.footerFrame);
+      if (summaryFrame === undefined) {
+        return { cells: undefined };
+      }
```
[the remark this answers is not in the document]
Comment (scottlepp) [reply public/app/plugins/panel/table/TablePanel.tsx]: We've hit the scenario where the dataset is too large with Oracle and MongoDB...and I'm sure other datasources.  Yes we need paging/infinite scrolling also.

File: packages/grafana-ui/src/components/Table/FooterRow.tsx [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+  }
+
+  return (
+    <table
+      style={{
+        position: 'absolute',
+        width: totalColumnsWidth ? `${totalColumnsWidth}px` : '100%',
+        bottom: '0px',
+      }}
+    >
```
Comment (dprokop) [inline packages/grafana-ui/src/components/Table/FooterRow.tsx:30]: It's just a nit, but what about using divs with roles to keep things consistent with how the entire Table component is built?

File: packages/grafana-ui/src/components/Table/Table.tsx [hunk 3]
```diff
@@ -233,44 +221,16 @@ export const Table: FC<Props> = memo((props: Props) => {
               No data
             </div>
           )}
+          <FooterRow
+            footer={footerValues}
+            footerGroups={footerGroups}
+            totalColumnsWidth={totalColumnsWidth}
+            height={footerHeight}
+          />
```
Comment (dprokop) [inline packages/grafana-ui/src/components/Table/Table.tsx:229]: Let's render only if footerValues are defined

File: public/app/plugins/panel/table/TablePanel.tsx [hunk 4]
```diff
@@ -91,21 +114,188 @@ export class TablePanel extends Component<Props> {
...
+
+  calculateSummaryCell(i: number, functionName: string, f: Field, showLabel: boolean) {
+    if (functionName === '') {
+      return undefined;
+    }
+
+    if (functionName && functionName !== 'avg') {
+      const kv = this as KeyValue;
+      const func: Function = kv[functionName];
+      const val = this.calculate(f.values, func);
```
Comment (dprokop) [inline public/app/plugins/panel/table/TablePanel.tsx:242]: Think using the getFieldDisplayValues should do the same job especially that the calcs may already be pre-calculated in the field.state. Similar thing is done i.e. in the Stat panel https://github.com/grafana/grafana/blob/073c84179fa0142732e10e21e87a0e5c74ddfd6b/public/app/plugins/panel/singlestat2/SingleStatPanel.tsx#L29

The formatting below should also be covered by getFieldDisplayValues as it takes field config into consideration.

File: public/app/plugins/panel/table/models.gen.ts [hunk 5]
```diff
@@ -15,6 +15,15 @@ export interface PanelOptions {
   frameIndex: number;
   showHeader: boolean;
   sortBy?: TableSortByFieldState[];
+  footerMode?: 'auto' | 'summary' | 'frame' | 'none';
+  footerFrame?: number;
+  footerSummary?: TableFooterSummary; // TODO: should be array (options builder is limited)
+  footerLabelMode?: 'auto' | 'first' | 'inline';
```
Comment (dprokop) [inline public/app/plugins/panel/table/models.gen.ts:21]: Can you make the modes enums?

File: public/app/plugins/panel/table/module.tsx [hunk 6]
```diff
@@ -75,10 +75,82 @@ export const plugin = new PanelPlugin<PanelOptions, PanelFieldConfig>(TablePanel
...
+          options: [
+            { value: 'min', label: 'Min' },
+            { value: 'max', label: 'Max' },
+            { value: 'sum', label: 'Sum' },
+            { value: 'avg', label: 'Avg' },
+          ],
+        },
+        defaultValue: 'sum',
+        showIf: (cfg) => cfg.footerMode === 'summary',
+      })
```
Comment (dprokop) [inline public/app/plugins/panel/table/module.tsx:114]: You can use std stats editor for this maybe: https://github.com/grafana/grafana/blob/53630b5f13095eea309c0560b6b100f6b1b8295c/public/app/plugins/panel/stat/types.ts#L79

File: packages/grafana-ui/src/components/Table/FooterRow.tsx (hunk 2 above)
Comment (scottlepp) [reply packages/grafana-ui/src/components/Table/FooterRow.tsx]: That was in a separate pr that is now merged to main. It was a div following the way it was done initially, though @hugohaggmark asked for more semantic html.  I didn't want to change everything so I only changed the footer.  Using divs with roles seems a bit better for consistency.  @hugohaggmark?

Comment (hugohaggmark) [reply packages/grafana-ui/src/components/Table/FooterRow.tsx]: Sure, I have no strong opinions as long as we follow a11y guidelines.

File: public/app/plugins/panel/table/TablePanel.tsx [hunk 7]
```diff
@@ -91,21 +105,184 @@ export class TablePanel extends Component<Props> {
...
+    const summaryFrame = data.series.find((f) => f.name === options.footerFrame);
+    if (summaryFrame === undefined) {
+      return undefined;
+    }
+
+    // group cells by field for multi value cells
+    const cellsByField: KeyValue = {};
+    frame.fields.forEach((f, i) => {
+      for (const sf of summaryFrame!.fields) {
+        const referenceField: string = sf.config?.custom?.['referenceField'];
```
Comment (scottlepp) [inline public/app/plugins/panel/table/TablePanel.tsx:213]: @ryantxu this is currently what I'm using to link the totals field/column to the details field/column.  If this is ok, seems we don't need to select a frame for the footer, just use the frame with this custom meta.

File: public/app/plugins/panel/table/models.gen.ts [hunk 8]
```diff
@@ -4,23 +4,48 @@
...
 export interface PanelOptions {
   frameIndex: number;
   showHeader: boolean;
   showTypeIcons?: boolean;
   sortBy?: TableSortByFieldState[];
+  footerMode?: FooterMode;
+  footerFrame?: number;
+  footerSummary?: TableFooterSummary; // TODO: should be array (options builder is limited)
+  footerLabelMode?: FooterLabelMode;
+  reduceOptions?: ReduceDataOptions;
```
Comment (ryantxu) [inline public/app/plugins/panel/table/models.gen.ts:37]: isn't the reducer part of the `footerSummary`?

File: public/app/plugins/panel/table/module.tsx [hunk 9]
```diff
@@ -75,10 +81,76 @@ export const plugin = new PanelPlugin<PanelOptions, PanelFieldConfig>(TablePanel
...
+        path: 'footerFrame',
+        name: 'Frame',
+        description: 'Select a Frame to use as the summary',
+        settings: {
+          options: [],
+          getOptions: async (context: FieldOverrideContext) => {
+            if (context && context.data && context.data.length > 1) {
+              const frames = context.data.filter((_, i) => i > 0);
+              return frames.map((f) => ({ value: f.name, label: f.name }));
+            }
```
Comment (ryantxu) [inline public/app/plugins/panel/table/module.tsx:150]:
```suggestion
            if (context?.data?.length) {
              return context.data.filter((_, i) => i > 0).map((f) => ({ value: f.name, label: f.name }));
            }
```

File: public/app/plugins/panel/table/module.tsx [hunk 10]
```diff
@@ -75,10 +81,76 @@ export const plugin = new PanelPlugin<PanelOptions, PanelFieldConfig>(TablePanel
...
+            // { label: 'auto', value: 'auto' },  // TODO: when frame meta supports this
+            { label: 'Summary', value: 'summary' },
+            { label: 'Provided', value: 'frame' },
+          ],
+        },
+        defaultValue: 'none',
+      })
+      .addCustomEditor({
+        category: ['Table footer'],
+        id: 'reduceOptions.calcs',
```
Comment (ryantxu) [inline public/app/plugins/panel/table/module.tsx:107]: should this be on: footerSummary.reducer?

File: public/app/plugins/panel/table/TablePanel.tsx (hunk 7 above)
Comment (ryantxu) [reply public/app/plugins/panel/table/TablePanel.tsx:214]: I don't really get what this is for -- is this to give us the same formatting?

Is it ever something different than the corresponding column?

Comment (scottlepp) [reply public/app/plugins/panel/table/TablePanel.tsx:214]: In salesforce the summary field name is "s!AMOUNT" (sum of AMOUNT) which is a reference to AMOUNT.  Could also have a field called a!AMOUNT (avg of AMOUNT).  If I strip the prefix I may have 2 fields with the same name.

File: public/app/plugins/panel/table/models.gen.ts (hunk 8 above)
Comment (scottlepp) [reply public/app/plugins/panel/table/models.gen.ts:37]: this is referenced by the stats editor that @dprokop suggested

File: public/app/plugins/panel/table/module.tsx (hunk 10 above)
Comment (scottlepp) [reply public/app/plugins/panel/table/module.tsx:107]: i wasn't sure what id did.  it sets a reference to the model?

File: public/app/plugins/panel/table/TablePanel.tsx (hunk 7 above)
Comment (scottlepp) [reply public/app/plugins/panel/table/TablePanel.tsx:214]: Summary Frame:
```
| Sum of Amount | Avg of Amount | 
----------------------------------
        500            200                      
```

Footer ( 2 fields from the summary frame reference one field in the detail frame and are stacked)
```
| Amount |
----------
 sum: 500
 avg: 200

```

File: public/app/plugins/panel/table/models.gen.ts (hunk 8 above)
Comment (scottlepp) [reply public/app/plugins/panel/table/models.gen.ts]: updated.  just needs to be ReduceDataOptions type when calling "getFieldDisplayValues":

```
    return getFieldDisplayValues({
      fieldConfig,
      reduceOptions: options.footerSummary?.reduceOptions || { calcs: [] },
      replaceVariables,
      theme: config.theme2,
      data: [data.series[currentIndex]],
      timeZone,
    });
```

File: public/app/plugins/panel/table/module.tsx (hunk 10 above)
Comment (scottlepp) [reply public/app/plugins/panel/table/module.tsx]: updated

**Comment by scottlepp** [comment]
@torkelo simplified/limited this and updated the comment with a recent screenshot and what it does now.  Can you please review again?  Thanks.

**Review by dprokop — COMMENTED** [review]
@scottlepp tested this and it looks really good. One thing I have noticed is that the Show footer option in the Table panel has no effect:

Kapture 2021-09-10 at 15 33 59

**Comment by scottlepp** [comment]
@dprokop just fixed that

**Comment by torkelo** [comment]
@scottlepp testing PR now

**Review by torkelo — COMMENTED** [review]
Looks good! 

Do we really need a new category for these options? Wondering if they cannot just live in the main Table category. There we have the `Show header` option and it feels strange that `Show footer` lives in a seperate category. We just need to add `Footer` prefix to the other options (ie `Footer calculation`, `Footer fields` ) 

Another thing missing is an update to the table test dashboard, it needs 1-2 panels that tests this feature. 
http://localhost:3000/d/U_bZIMRMk/panel-tests-react-table?orgId=1 

I think it's this: https://github.com/grafana/grafana/blob/main/devenv/dev-dashboards/panel-table/table_tests_new.json

**Comment by scottlepp** [comment]
@torkelo eventually we want more features on the footer:
* allow multiple reducers ( would make the section much longer )
* allow footers to be show from another frame/backend calculation ( hopefully )

The footer section will grow.

**Comment by scottlepp** [comment]
@torkelo just removed the footer category.  We can split into 2 if/when we add the new features.

File: public/app/plugins/panel/table/module.tsx [hunk 11]
```diff
@@ -75,10 +82,52 @@ export const plugin = new PanelPlugin<PanelOptions, PanelFieldConfig>(TablePanel
...
+      })
+      .addMultiSelect({
+        path: 'footer.fields',
+        name: 'Fields',
+        description: 'Select the fields that should be calculated',
+        settings: {
+          allowCustomValue: false,
+          options: [],
+          placeholder: 'All Numeric Fields',
+          getOptions: async (context: FieldOverrideContext) => {
```
Comment (torkelo) [inline public/app/plugins/panel/table/module.tsx:115]: So we have a addFieldNamePicker but I see it's missing a multi-select option, wonder if we should add that to that option or have some other built in multi fields selector.  throughts @ryantxu ?

File: public/app/plugins/panel/table/module.tsx [hunk 12]
```diff
@@ -75,10 +82,52 @@ export const plugin = new PanelPlugin<PanelOptions, PanelFieldConfig>(TablePanel
...
+        id: 'footer.reducer',
+        path: 'footer.reducer',
+        name: 'Calculation',
+        description: 'Choose a reducer function / calculation',
+        editor: standardEditorsRegistry.get('stats-picker').editor as any,
+        defaultValue: [ReducerID.sum],
+        showIf: (cfg) => cfg.showFooter,
+      })
+      .addMultiSelect({
+        path: 'footer.fields',
```
Comment (torkelo) [inline public/app/plugins/panel/table/module.tsx:108]: Nitpick: I am torn on the mix of nested and non nested footer options (showFooter and footer.reducer), not sure what we should prefer/start with.

Comment (torkelo) [inline public/app/plugins/panel/table/module.tsx:108]: Nitpick: I am torn on the mix of nested and non nested footer options (showFooter and footer.reducer), not sure what we should prefer/start with.

**Review by torkelo — COMMENTED** [review]
This looks good, but there is one issue I see now, when I select Total in the Calculations (the default) , it says "SUM" in the visualisation, should be the same name (the name of the reducer is Total, I know it's confusing

**Review by torkelo — APPROVED** [review]

**Pull request merged**